pkg/catalogv2/git/git.go

```diff
@@ -1,480 +1,395 @@
-	}
-	return g, g.setCredential(opts.Credential)
-}
+	// gogit packages
+	gogit "github.com/go-git/go-git/v5"
+	config "github.com/go-git/go-git/v5/config"
+	plumbing "github.com/go-git/go-git/v5/plumbing"
+	transport "github.com/go-git/go-git/v5/plumbing/transport"
+	plumbingHTTP "github.com/go-git/go-git/v5/plumbing/transport/http"
+	plumbingSSH "github.com/go-git/go-git/v5/plumbing/transport/ssh"
+)
 
-type git struct {
+// repository holds the config of a git repository and the repo instance.
+type Repository struct {
 	URL               string
 	Directory         string
+	username          string
 	password          string
 	agent             *agent.Agent
 	caBundle          []byte
 	insecureTLSVerify bool
 	secret            *corev1.Secret
-	headers           map[string]string
 	knownHosts        []byte
+	repoGogit         *gogit.Repository
+	auth              transport.AuthMethod
+	cloneOpts         *gogit.CloneOptions
+	fetchOpts         *gogit.FetchOptions
+	listOpts          *gogit.ListOptions
+	resetOpts         *gogit.ResetOptions
 }
 
-// LsRemote runs ls-remote on git repo and returns the HEAD commit SHA
-func (g *git) LsRemote(branch string, commit string) (string, error) {
-	if changed, err := g.remoteSHAChanged(branch, commit); err != nil || !changed {
-		return commit, err
-	}
-
-	output := &bytes.Buffer{}
-	if err := g.gitCmd(output, "ls-remote", "--", g.URL, formatRefForBranch(branch)); err != nil {
-		return "", err
-	}
-
-	var lines []string
-	s := bufio.NewScanner(output)
-	for s.Scan() {
-		lines = append(lines, s.Text())
-	}
-
-	return firstField(lines, fmt.Sprintf("no commit for branch: %s", branch))
-}
-
-// Head runs git clone on directory(if not exist), reset dirty content and return the HEAD commit
-func (g *git) Head(branch string) (string, error) {
-	if err := g.clone(branch); err != nil {
-		return "", err
-	}
-
-	if err := g.reset("HEAD"); err != nil {
-		return "", err
+// BuildRepoConfig constructs and returns a new repository object for the given repository.
+// It requires a secret for authentication, the namespace, the name of the repository,
+// the gitURL, a flag indicating if TLS verification should be skipped, and a CA bundle for SSL.
+// If the Git URL uses the SSH protocol, it checks if the URL is a valid SSH URL.
+// If the Git URL uses HTTP or HTTPS, it parses and verifies the URL.
+// It then constructs a directory path for the git repository.
+// If a CA bundle is provided, it converts the CA bundle from DER to PEM format, since Git requires PEM format.
+// In this case, insecureSkipTLS is set to false since a CA bundle is provided for secure communication.
+// Finally, it returns a new git object configured with these settings,
+// or an error if any step in this process fails.
+func BuildRepoConfig(secret *corev1.Secret, namespace, name, gitURL string, insecureSkipTLS bool, caBundle []byte) (*Repository, error) {
+	isGitSSH, err := isGitSSH(gitURL)
+	if err != nil {
```

Comment:
An error here means that `regexp.Compile` failed to parse the regex, not that the url itself is invalid, can't we try to check if it's a valid http url regardless of an error here?

Reply:
This checks if it is a valid SSH URL . 
If it is not it will not throw an error. 
It just keeps going, actually, I have never seen a case where this breaks and throws an error.

But do you mean, to check if it is a valid SSH URL and then check again if it is a valid HTTP(S) URL?

Reply:
Yes, I meant to check both. You are returning an error here, but the only error that the `isGitSSH` can return is an invalid regex, so that doesn't necessarily means that the URL is invalid